WordPress/Classes/Services/PostCoordinator.swift

```diff
@@ -0,0 +1,70 @@
+import Foundation
+
+class PostCoordinator: NSObject {
+
+    @objc static let shared = PostCoordinator()
+
+    private(set) var backgroundContext: NSManagedObjectContext = {
+        let context = ContextManager.sharedInstance().newDerivedContext()
+        context.automaticallyMergesChangesFromParent = true
+        return context
+    }()
+
+    private let mainContext = ContextManager.sharedInstance().mainContext
+
+    private let queue = DispatchQueue(label: "org.wordpress.postcoordinator")
+
+    /// Saves the post to both the local database and the server if available.
+    /// If media is still uploading it keeps track of the ongoing media operations and updates the post content when they finish
+    ///
+    /// - Parameter post: the post to save
+    func save(post: AbstractPost) {
+        let mediaCoordinator = MediaCoordinator.shared
+
+        if mediaCoordinator.isUploadingMedia(for: post) {
+            mediaCoordinator.addObserver({ (media, state) in
+                switch state {
+                case .ended:
+                    self.updateReferences(to: media, in: post)
+                    // Let's check if media uploading is still going, if all finished with success then we can upload the post
+                    if !mediaCoordinator.isUploadingMedia(for: post) {
+                        self.upload(post: post)
+                    }
+                default:
+                    print("Post Coordinator -> Media state: \(state)")
+                }
+            }, forMediaFor: post)
+            return
+        }
+
+        upload(post: post)
+    }
+
+    private func upload(post: AbstractPost) {
+        let postService = PostService(managedObjectContext: mainContext)
+        postService.uploadPost(post, success: { uploadedPost in
+            print("Post Coordinator -> upload succesfull: \(String(describing: uploadedPost.content))")
+        }) { error in
```

Comment:
Let's keep the last parameter name here, as it makes things a little clearer -> `}, failure: { ...`